### PR TITLE
fix: UNIFY-676 browsers should be configurable in setupNodeEvents

### PR DESCRIPTION
**Pull request by tgriesser**
- Closes UNIFY-676

Adds a system test to show that filtering browsers is working properly & `$ExpectType` assertion for `dtslint` to assert the correct type, @JessicaSachs' original issue

### CODE REVIEW

**Comment by tbiethman** [comment]
We just need to interpolate the majorVersion now: https://github.com/cypress-io/cypress/blob/a015397f577313b9ae09c86fbeea11cc067ed724/packages/server/lib/browsers/chrome.ts#L407

**Comment by tgriesser** [comment]
I had updated the type because it seemed to be wrong but who knows, the types in the server layer are pretty tough to follow. Didn't need to address it as part of this PR.

**Review by tbiethman — DISMISSED** [review]

**Review by elevatebart — DISMISSED** [review]

**Review by JessicaSachs — COMMENTED** [review]
Doesn't work, Tim and I DM'd about it already.

**Review by JessicaSachs — CHANGES_REQUESTED** [review]
Doesn't work, Tim and I DM'd about it already.

image

**Review by elevatebart — CHANGES_REQUESTED** [review]
Would this mean that now `config.browser` is available in the types of `setupNodeEvents()`?

The value will always be null and setting it should have no effect. 

How far off am I ?

**Comment by lmiller1990** [comment]
Seems fine to me, I followed https://docs.cypress.io/guides/guides/launching-browsers#Customize-available-browsers and got this:
 
image

**Review by lmiller1990 — DISMISSED** [review]

**Comment by lmiller1990** [comment]
Alrighty I got this green, let's get one more test/approval. The types are working fine for me, too, from what I can see. I had to specify `headless` and `isHeaded` which seems weird, but fixing the types in the entirely of Cypress is a huge task.

File: cli/types/tests/plugins-config.ts [hunk 1]
```diff
@@ -16,6 +16,8 @@ const pluginConfig2: Cypress.PluginConfig = (on, config) => {
   config.version // $ExpectType: string
   config.testingType // $ExpectType: TestingType
 
+  config.browsers // $ExpectType: false
```
Comment (tgriesser) [inline cli/types/tests/plugins-config.ts:19]: @lmiller1990 this is what I was most recently looking into. This was set as a failing test but it doesn't seem to be triggered. Looking to see where we regressed in terms of running the test on these

**Comment by tgriesser** [comment]
@lmiller1990 I'm not sure why all of these changes were even needed. I wasn't sure if the types were correct

**Comment by tgriesser** [comment]
Changing to draft to investigate why the `$ExpectTypre` types aren't being checked

**Comment by tgriesser** [comment]
@elevatebart:

> Would this mean that now config.browser is available in the types of setupNodeEvents()?

Nah, it's using a `Partial` so it'll be `Browser | undefined`

**Review by lmiller1990 — APPROVED** [review]
Seems fine to me! I tried it as mentioned in a previous review and it worked okay, including types.

**Review by elevatebart — APPROVED** [review]
@tgriesser I beg to disagree, the problem is still here:
The code below seems to be valid typed configFile. 

```ts
import { defineConfig } from 'cypress'

export default defineConfig({
  e2e: {
    setupNodeEvents (on, config) {
      config.browser = 'chrome'
    },
  },
})
```

What I am saying is that the types are not correct since they contain more than the actual object. Instead of helping the types we have here could confuse users. 
I prefer that we do not fix it in this PR though.

**Comment by lmiller1990** [comment]
@elevatebart what's an example of a property on config that's not actually there? I think since it's a `Partial` everything is going to be nullable, which seems fine - users will basically need to assume any of the properties they are accessing are nullable and do the required defensive checks.

**Pull request merged**